pkg/assembler/backends/testing/hashEqual.go

```diff
@@ -17,113 +17,205 @@ package testing
 	}
-
-	newHashEqual := &model.HashEqual{
-		Justification: justification,
-		Artifacts:     artifacts,
-		Origin:        origin,
-		Collector:     collector,
+	a, ok := o.(*hashEqualStruct)
+	if !ok {
+		return nil, errors.New("not a hashEqual")
 	}
-	c.hashEquals = append(c.hashEquals, newHashEqual)
-	return newHashEqual
+	return a, nil
 }
 
+// Ingest HashEqual
 func (c *demoClient) IngestHashEqual(ctx context.Context, artifact model.ArtifactInputSpec, equalArtifact model.ArtifactInputSpec, hashEqual model.HashEqualInputSpec) (*model.HashEqual, error) {
 
-	collectedArt, err := c.Artifacts(ctx, &model.ArtifactSpec{Algorithm: &artifact.Algorithm, Digest: &artifact.Digest})
+	aInt1, err := c.artifactByKey(artifact.Algorithm, artifact.Digest)
 	if err != nil {
-		return nil, err
-	}
-	if len(collectedArt) != 1 {
-		return nil, gqlerror.Errorf(
-			"IngestHashEqual :: multiple artifacts found")
+		return nil, gqlerror.Errorf("IngestHashEqual :: Artifact not found")
 	}
-
-	collectedEqualArt, err := c.Artifacts(ctx, &model.ArtifactSpec{Algorithm: &equalArtifact.Algorithm, Digest: &equalArtifact.Digest})
+	aInt2, err := c.artifactByKey(equalArtifact.Algorithm, equalArtifact.Digest)
 	if err != nil {
-		return nil, err
+		return nil, gqlerror.Errorf("IngestHashEqual :: Artifact not found")
 	}
-	if len(collectedEqualArt) != 1 {
-		return nil, gqlerror.Errorf(
-			"IngestHashEqual :: multiple artifacts found")
+	artIDs := []uint32{aInt1.id, aInt2.id}
+	sort.Slice(artIDs, func(i, j int) bool { return artIDs[i] < artIDs[j] })
+
+	// Search backedges for existing.
+	searchHEs := slices.Clone(aInt1.hashEquals)
+	searchHEs = append(searchHEs, aInt2.hashEquals...)
+
+	for _, he := range searchHEs {
+		h, err := c.hashEqualByID(he)
+		if err != nil {
+			return nil, gqlerror.Errorf(
+				"IngestHashEqual :: Bad hashEqual id stored on existing artifact: %s", err)
+		}
+		if h.justification == hashEqual.Justification &&
+			h.origin == hashEqual.Origin &&
+			h.collector == hashEqual.Collector &&
+			slices.Equal(h.artifacts, artIDs) {
+			return c.convHashEqual(h), nil
+		}
 	}
 
-	return c.registerHashEqual(
-		[]*model.Artifact{collectedArt[0], collectedEqualArt[0]},
-		hashEqual.Justification,
-		hashEqual.Origin,
-		hashEqual.Collector), nil
+	he := &hashEqualStruct{
+		id:            c.getNextID(),
+		artifacts:     artIDs,
+		justification: hashEqual.Justification,
+		origin:        hashEqual.Origin,
+		collector:     hashEqual.Collector,
+	}
+	c.index[he.id] = he
+	c.hashEquals = append(c.hashEquals, he)
+	aInt1.hashEquals = append(aInt1.hashEquals, he.id)
+	aInt2.hashEquals = append(aInt2.hashEquals, he.id)
+
+	return c.convHashEqual(he), nil
 }
 
-// Query HashEqual
+func (c *demoClient) matchArtifacts(filter []*model.ArtifactSpec, value []uint32) bool {
+	val := slices.Clone(value)
+	var matchID []uint32
+	var matchPartial []*model.ArtifactSpec
+	for _, aSpec := range filter {
+		a, _ := c.artifactExact(aSpec)
+		// drop error here if ID is bad
```

Comment:
not sure what you mean by the comment - can you explain further why we skip error checks here? if it shouldn't happen should we have a check for and panic if it errors?

Reply:
Yes, overall I'm dropping errors here and there. I believe all are from looking up IDs in the index that are already stored on existing nodes. All these should never error. Not sure if we should catch those and explicitly panic, or find a way to bubble the error up and return a gql "internal error" or something similar.  Currently they will probably panic on a pointer dereference later on.

Reply:
Added #606 606